### PR TITLE
Added handleSubmit

### PR DESCRIPTION
**Pull request by PorterK**

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,3 +1,3 @@
 {
-  "presets": ["es2015", "react"]
+  "presets": ["es2015", "react", "stage-2"]
```
Comment (NullVoxPopuli) [inline .babelrc]: Which features are in stage 2 that you're using?

File: package.json [hunk 2]
```diff
@@ -30,10 +30,13 @@
     "babel-core": "^6.24.1",
     "babel-preset-es2015": "^6.24.1",
     "babel-preset-react": "^6.24.1",
+    "babel-preset-stage-2": "^6.24.1",
     "babel-require": "^1.0.1",
     "deepmerge": "^1.3.2",
     "enzyme": "^2.8.2",
     "expect": "^1.20.2",
+    "jsdom": "^10.1.0",
+    "jsdom-global": "^3.0.2",
```
Comment (NullVoxPopuli) [inline package.json]: Jsdom?

File: src/index.js [hunk 3]
```diff
@@ -34,6 +34,26 @@ export function toggleCreator(context) {
...
+    const form = e.target;
+    let values = {};
+
+    let children = [];
+
+    children.push(form.querySelectorAll('input, select, textarea'));
+
+    for(let element of children[0]){
+
+      values = { ...values, [element.name]: element.value };
```
Comment (NullVoxPopuli) [inline src/index.js]: Can you add a comment here about what values will look like? (Just an example or something)

File: README.md [hunk 4]
```diff
@@ -139,6 +139,28 @@ export default connect(
...
+this.submit = (values) => {
+  dispatch(actions.login(values.username, values.password));
+}
+```
+```js
+// in render method
+const { submit } = this;
+
+return(
+  <form onSubmit={submit}>
```
Comment (NullVoxPopuli) [inline README.md]: Does handleSubmit need to be mentioned?

File: tests/index.test.js [hunk 5]
```diff
@@ -64,6 +66,34 @@ class ComponentStub {
...
+
+  render(){
+    return(
+      <form onSubmit={handleSumbit(Actions.submit)}>
+        <input name='testInput' type='text' defaultValue='value' />
+        <div>
+          <input name='nestedInput' type='text' defaultValue='nestedValue' />
+        </div>
+        <select name='testSelect'>
+          <option value='test1'>Test 1</option>
```
Comment (NullVoxPopuli) [inline tests/index.test.js]: Can you add two more options and have the middle selected? Right now it's just picking the first thing

File: tests/index.test.js [hunk 6]
```diff
@@ -64,6 +66,34 @@ class ComponentStub {
...
+class FormComponent extends Component{
+  constructor(props){
+    super(props);
+
+    this.state = {};
+  }
+
+  render(){
+    return(
+      <form onSubmit={handleSumbit(Actions.submit)}>
```
Comment (NullVoxPopuli) [inline tests/index.test.js]: Can you add a couple check boxes, and radios? (Just for thorough Ness)

File: tests/index.test.js [hunk 7]
```diff
@@ -191,3 +221,38 @@ describe('toggleCreator', () => {
...
+
+      wrapper.find('button').get(0).click();
+    });
+
+    afterEach(() => {
+      expect.restoreSpies();
+    })
+
+    it('returns a value for an input', () => {
+      expect(Actions.submit.calls[0].arguments[0].testInput).toEqual('value');
```
Comment (NullVoxPopuli) [inline tests/index.test.js]: Since each of these tests is getting daata from the same place, it could be extracted :)

File: src/index.js [hunk 8]
```diff
@@ -34,6 +34,26 @@ export function toggleCreator(context) {
...
+export function handleSumbit(func){
+  return (e) => {
+    e.preventDefault();
+
+    const form = e.target;
+    let values = {};
+
+    let children = [];
+
+    children.push(form.querySelectorAll('input, select, textarea'));
```
Comment (NullVoxPopuli) [inline src/index.js]: I still wonder if using elements is cleaner. With the query selector, we'd need to update it if we had other things to collect info from.  But elements would do that for us, I think? Unless I'm mistaken.

In slack, you said elements includes the button? I don't think that's a problem. 

Though, I wonder if there is a way to do .elements&[name] or something like​ that. Cause we can only care about named things, yeah?

File: .babelrc (hunk 1 above)
Comment (PorterK) [inline .babelrc]: The babel spread operator is stage 2

File: src/index.js (hunk 8 above)
Comment (PorterK) [inline src/index.js]: Elements is the exact same list

https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/elements

I wasn't able to pull out buttons in a test right now, hmm.

File: src/index.js (hunk 3 above)
Comment (PorterK) [inline src/index.js]: Yeah, no problem

File: README.md (hunk 4 above)
Comment (PorterK) [inline README.md]: yes lol, sorry it was late-- will fix

File: tests/index.test.js (hunk 5 above)
Comment (PorterK) [inline tests/index.test.js]: Yep

File: tests/index.test.js (hunk 7 above)
Comment (PorterK) [inline tests/index.test.js]: will do

File: src/index.js (hunk 8 above)
Comment (NullVoxPopuli) [inline src/index.js]: most excellent news! lets use elements then (native apis are gettin good) woo

Ya don't really need to pull out the buttons. 
I'd do:
```
_.compact(
  elements
    .map(e => e.hasAttribute('name')))
    .map(e => ({ [e.name]: e.value }))
```
or something similar (I haven't run that, so it probably doesn't work. lol)

File: README.md (hunk 4 above)
Comment (NullVoxPopuli) [inline README.md]: np, bro

File: .babelrc (hunk 1 above)
Comment (NullVoxPopuli) [inline .babelrc]: noice

File: src/index.js [hunk 9]
```diff
@@ -34,6 +34,40 @@ export function toggleCreator(context) {
...
+    children.push(form.elements);
+
+    for(let element of children[0]){
+      // Values will look like objectName: value
+
+      switch(element.type){
+        case 'checkbox':
+          values = { ...values, [element.name]: element.checked };
+          break;
+        // RadioNodeList doesn't work with every browser, check for checked instead.
```
Comment (NullVoxPopuli) [inline src/index.js]: figures

File: src/index.js (hunk 8 above)
Comment (PorterK) [inline src/index.js]: I tested it again this morning and I wansn't getting buttons... 
worse case scenario, the values object will have a hanging `"":""`

File: package.json (hunk 2 above)
Comment (PorterK) [inline package.json]: jsdom is required to use `mount()` afaik

(maybe not required, but easiest solution anyways)

File: src/index.js [hunk 10]
```diff
@@ -34,6 +34,40 @@ export function toggleCreator(context) {
...
+    e.preventDefault();
+
+    const form = e.target;
+    let values = {};
+
+    let children = [];
+
+    children.push(form.elements);
+
+    for(let element of children[0]){
```
Comment (NullVoxPopuli) [inline src/index.js]: why bother with a children array if just accessing the first element of it? 

you could do `form.elements.forEach(...)`?

File: tests/index.test.js [hunk 11]
```diff
@@ -64,6 +66,42 @@ class ComponentStub {
...
+  render(){
+    return(
+      <form onSubmit={handleSumbit(Actions.submit)}>
+        <input id='testInput' name='testInput' type='text' defaultValue='value' />
+        <div>
+          <input id='nestedInput' name='nestedInput' type='text' defaultValue='nestedValue' />
+        </div>
+        <select id='testSelect' name='testSelect'>
+          <option value='test1'>Test 1</option>
+          <option value='test2'>Test 2</option>
```
Comment (NullVoxPopuli) [inline tests/index.test.js]: does this work if the second one is selected?

File: tests/index.test.js [hunk 12]
```diff
@@ -191,3 +229,111 @@ describe('toggleCreator', () => {
...
+
+
+    beforeEach(() => {
+      wrapper = mount(<FormComponent />);
+      submit = expect.spyOn(Actions, 'submit');
+
+      wrapper.find('#testInput').node.value = 'ChangedTestInput';
+      wrapper.find('#nestedInput').node.value = 'ChangedNestedInput';
+      wrapper.find('#testSelect').node.value = 'test2';
+      wrapper.find('#testTextArea').node.value='MutatedTest';
```
Comment (NullVoxPopuli) [inline tests/index.test.js]: linting here

File: package.json (hunk 2 above)
Comment (NullVoxPopuli) [inline package.json]: huh

File: tests/index.test.js (hunk 11 above)
Comment (PorterK) [inline tests/index.test.js]: Yes, and I've written a test for it.

File: tests/index.test.js (hunk 12 above)
Comment (PorterK) [inline tests/index.test.js]: My linter isn't working out of the box with this module.. hm

File: tests/index.test.js (hunk 11 above)
Comment (NullVoxPopuli) [inline tests/index.test.js]: :+1:

File: tests/index.test.js (hunk 12 above)
Comment (NullVoxPopuli) [inline tests/index.test.js]: huh. well, I guess if there is no error, it can't be enforced. lol.

File: src/index.js (hunk 10 above)
Comment (NullVoxPopuli) [inline src/index.js]: @PorterK ^

**Comment by NullVoxPopuli** [comment]
@PorterK So, I was curious what you were up against with this. I made some minor cleanup changes, but I broke one of the radio tests -- any idea what happened there?

**Comment by NullVoxPopuli** [comment]
Also, I need to add an eslintrc. Codeclimate so mad

**Comment by NullVoxPopuli** [comment]
Got all the tests passing.

File: src/index.js [hunk 13]
```diff
@@ -34,6 +34,44 @@ export function toggleCreator(context) {
...
+    e.preventDefault();
+
+    const form = e.target;
+    const elements = form.elements;
+    const numElements = elements.length;
+    let values = {};
+
+    for (let i = 0; i < numElements; i++) {
+      const input = elements[i];
+      const value = valueOfInput(input);
```
Comment (NullVoxPopuli) [inline src/index.js]: extracted the logic for ease of readability.

**Pull request merged**